service/worker/batcher/workflow.go

```diff
@@ -20,7 +20,328 @@
+	}
+	switch params.BatchType {
+	case BatchTypeReset:
+		fallthrough
+	default:
+		return BatchParams{}, fmt.Errorf("not supported batch type: %v", params.BatchType)
+	case BatchTypeTerminate:
+
+	}
+	return params, nil
+}
+
+func BatchActivity(ctx context.Context, batchParams BatchParams) error {
+	batcher := ctx.Value(batcherContextKey).(*Batcher)
+
+	hbd := HeartBeatDetails{}
+	startOver := true
+	if activity.HasHeartbeatDetails(ctx) {
+		if err := activity.GetHeartbeatDetails(ctx, &hbd); err == nil {
+			startOver = false
+		} else {
+			//TODO error metrics
+			getActivityLogger(ctx).Error("Failed to recover from last heartbeat, start over from beginning", tag.Error(err))
+		}
+	}
+
+	if startOver {
+		resp, err := batcher.svcClient.CountWorkflowExecutions(ctx, &shared.CountWorkflowExecutionsRequest{
+			Domain: common.StringPtr(batchParams.DomainName),
+			Query:  common.StringPtr(batchParams.QueryCondition),
+		})
+		if err != nil {
+			return err
+		}
+		hbd.TotalEstimate = resp.GetCount()
+	}
+	rateLimiter := rate.NewLimiter(rate.Limit(batchParams.RPS), batchParams.RPS)
+	taskCh := make(chan taskDetail, pageSize)
+	var succCount, errCount int32
+	for i := 0; i < batchParams.Concurrency; i++ {
+		go startTaskProcessor(ctx, batchParams, taskCh, rateLimiter, &succCount, &errCount)
+	}
+
+	for {
+		// TODO https://github.com/uber/cadence/issues/2154
+		//  Need to improve scan concurrency because it will hold a ES resource for a long time.
+		//  we can't use list API because terminate / reset will mutate the result.
+		resp, err := batcher.svcClient.ScanWorkflowExecutions(ctx, &shared.ListWorkflowExecutionsRequest{
+			PageSize:      common.Int32Ptr(int32(pageSize)),
+			Domain:        common.StringPtr(batchParams.DomainName),
+			NextPageToken: hbd.PageToken,
+			Query:         common.StringPtr(batchParams.QueryCondition),
+		})
+		if err != nil {
+			return err
+		}
+
+		// clear all counters
+		atomic.StoreInt32(&succCount, 0)
+		atomic.StoreInt32(&errCount, 0)
+
+		// send all tasks
+		for _, wf := range resp.Executions {
+			taskCh <- taskDetail{
+				execution: *wf.Execution,
+				attempts:  0,
+				hbd:       hbd,
+			}
+		}
+		batchCount := len(resp.Executions)
+
+		// wait for counters indicate this batch is done
+	Loop:
+		for {
+			select {
+			case <-time.Tick(time.Second):
+				if int32(batchCount) == atomic.LoadInt32(&succCount)+atomic.LoadInt32(&errCount) {
+					break Loop
+				}
+			case <-ctx.Done():
+				return nil
+			}
+		}
+
+		hbd.CurrentPage++
+		hbd.PageToken = resp.NextPageToken
+		hbd.SuccessCount += atomic.LoadInt32(&succCount)
+		hbd.ErrorCount += atomic.LoadInt32(&errCount)
+		activity.RecordHeartbeat(ctx, hbd)
+
+		if len(hbd.PageToken) == 0 {
+			break
+		}
+	}
+
+	return nil
+}
+
+func startTaskProcessor(
+	ctx context.Context, batchParams BatchParams, taskCh chan taskDetail,
+	limiter *rate.Limiter, doneCount, errCount *int32) {
+
+	for {
+		select {
```

Comment:
Need to add done ctx or channel handing, otherwise this startTaskProcessor goroutine will leak. 
One leak case: goroutines start but no tasks produced in taskCh.

Reply:
done